### ISSUE TITLE
Result of `max` with `key=len` is always `Sized`

### ISSUE THREAD

**Issue by druzsan** [astral-sh/ty#2572]
### Summary

In the following code I search the longest element in a collection (I tried values of a dict and list):
```python
xs: list[list[str]] = [["foo", "bar"], ["baz", "qux", "quux"]]

x = max(xs, key=len)
reveal_type(x)
```

I expect `x` to be revealed as `list[str]`, but it is revealed as `Sized`.

Am I missing something?

Command executed:
```sh
ty check --error-on-warning
```

No other settings are applied to `ty`.

### Version

0.0.12

**Comment by MeGaGiGaGon** [comment]
This happens because the selected overload of `max` for this case is:
```py
@overload
def max(iterable: Iterable[_T], /, *, key: Callable[[_T], SupportsRichComparison]) -> _T: ...
```
And where ty is going wrong is in solving for `_T` - `xs` is `list[list[str]]`, but `len` is defined as:
```py
def len(obj: Sized, /) -> int: ...
```
Which is a `Callable[[Sized], SupportsRichComparison]`, so ty ends up assigning `Sized` to `_T` instead.
(And just for context, `Sized` is this):
```py
@runtime_checkable
class Sized(Protocol, metaclass=ABCMeta):
    @abstractmethod
    def __len__(self) -> int: ...
```
So in this case ty should prefer solving to the concrete class instead of the `Protocol`, but doesn't.

**Comment by carljm** [comment]
Another example that was reported that looks like it has the same root cause:

```py
s1: set[int] = {1, 2, 3}
s2: set[int] = {4, 5, 6, 7}
for idx in min(s1, s2, key=len):  # ty: Object of type `Sized` is not iterable
    print(idx)
```

**Comment by carljm** [comment]
Another example (reported by @oopus in #1714):

When sorting a list[str] with key=len, ty widens the element type to Sized, which then causes false positives when the result is used as str.

```py
from __future__ import annotations

import re

def build_regex(forms: list[str]) -> str:
    sorted_forms = sorted(forms, key=len)

    reveal_type(sorted_forms)  # ty infers list[Sized] (unexpected)
    # Any downstream use that requires `str` then fails:
    escaped = [re.escape(f) for f in sorted_forms]  # ty: expected str/bytes, got Sized
    return "|".join(escaped)
```

Expected:

    sorted_forms should be inferred as list[str] (same element type as the input iterable).

Actual:

    sorted_forms is inferred as list[Sized], losing the concrete element type and producing false positives.

**Comment by dcreager** [comment]
I opened https://github.com/astral-sh/ty/issues/2799 with a deeper explanation of what's going wrong with this issue (and several others).

**Comment by carljm** [comment]
I'm going to reopen this and make it a child issue of #2799, for symptom discoverability since we continue to get reports of this.
